### PR TITLE
Metal support for Swift

### PR DESCRIPTION
**Pull request by kchro3**
I have a working demo of using Metal w/ a Swift Mac app with these changes. Hopefully, this is a welcome contribution!

```
llama_model_loader: loaded meta data with 19 key-value pairs and 363 tensors from /Users/jeffhara/.cache/lm-studio/models/TheBloke/MythoMax-L2-Kimiko-v2-13B-GGUF/mythomax-l2-kimiko-v2-13b.Q4_K_M.gguf (version GGUF V2 (latest))
llama_model_loader: - tensor    0:                token_embd.weight q4_K     [  5120, 32000,     1,     1 ]
...B (+  400.00 MB per state)
...................................................................................................
llama_new_context_with_model: kv self size  =  400.00 MB
ggml_metal_init: allocating
2023-09-07 22:35:08.973604-0700 TestTypeaheadAI[60657:1063549] Metal GPU Frame Capture Enabled
2023-09-07 22:35:08.973931-0700 TestTypeaheadAI[60657:1063549] Metal API Validation Enabled
ggml_metal_init: found device: Apple M1 Pro
ggml_metal_init: picking default device: Apple M1 Pro
ggml_metal_init: loaded kernel_add                         0x600000fd0370 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_add_row                     0x600000fd0500 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_mul                         0x600000fd0690 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_mul_row                     0x600000fd0820 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_scale                       0x600000fd44b0 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_silu                        0x600000fd4640 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_relu                        0x600000fd47d0 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_gelu                        0x600000fee7b0 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_soft_max                    0x600000fee940 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_diag_mask_inf               0x600000feead0 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_get_rows_f16                0x600000feec60 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_get_rows_q4_0               0x600000feedf0 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_get_rows_q4_1               0x600000feef80 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_get_rows_q8_0               0x600000fd09b0 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_get_rows_q2_K               0x600000fd0b40 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_get_rows_q3_K               0x600000fd0cd0 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_get_rows_q4_K               0x600000fd0e60 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_get_rows_q5_K               0x600000fd0ff0 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_get_rows_q6_K               0x600000fd1180 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_rms_norm                    0x600000fd1310 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_norm                        0x600000fd14a0 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_mul_mat_f16_f32             0x600000fd1630 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_mul_mat_q4_0_f32            0x600000fd17c0 | th_max =  896 | th_width =   32
ggml_metal_init: loaded kernel_mul_mat_q4_1_f32            0x600000fd1950 | th_max =  896 | th_width =   32
ggml_metal_init: loaded kernel_mul_mat_q8_0_f32            0x600000fd1ae0 | th_max =  768 | th_width =   32
ggml_metal_init: loaded kernel_mul_mat_q2_K_f32            0x600000fd1c70 | th_max =  640 | th_width =   32
ggml_metal_init: loaded kernel_mul_mat_q3_K_f32            0x600000fd1e00 | th_max =  704 | th_width =   32
ggml_metal_init: loaded kernel_mul_mat_q4_K_f32            0x600000fd1f90 | th_max =  576 | th_width =   32
ggml_metal_init: loaded kernel_mul_mat_q5_K_f32            0x600000fd2120 | th_max =  576 | th_width =   32
ggml_metal_init: loaded kernel_mul_mat_q6_K_f32            0x600000fd22b0 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_mul_mm_f16_f32              0x600000fd2440 | th_max =  768 | th_width =   32
ggml_metal_init: loaded kernel_mul_mm_q4_0_f32             0x600000fd25d0 | th_max =  768 | th_width =   32
ggml_metal_init: loaded kernel_mul_mm_q8_0_f32             0x600000fd2760 | th_max =  768 | th_width =   32
ggml_metal_init: loaded kernel_mul_mm_q4_1_f32             0x600000fd28f0 | th_max =  768 | th_width =   32
ggml_metal_init: loaded kernel_mul_mm_q2_K_f32             0x600000fd2a80 | th_max =  768 | th_width =   32
ggml_metal_init: loaded kernel_mul_mm_q3_K_f32             0x600000fd2c10 | th_max =  768 | th_width =   32
ggml_metal_init: loaded kernel_mul_mm_q4_K_f32             0x600000fd2da0 | th_max =  768 | th_width =   32
ggml_metal_init: loaded kernel_mul_mm_q5_K_f32             0x600000fd2f30 | th_max =  704 | th_width =   32
ggml_metal_init: loaded kernel_mul_mm_q6_K_f32             0x600000fd30c0 | th_max =  704 | th_width =   32
ggml_metal_init: loaded kernel_rope                        0x600000fd3250 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_alibi_f32                   0x600000fd33e0 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_cpy_f32_f16                 0x600000fd3610 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_cpy_f32_f32                 0x600000fd37a0 | th_max = 1024 | th_width =   32
ggml_metal_init: loaded kernel_cpy_f16_f16                 0x600000fd3930 | th_max = 1024 | th_width =   32
ggml_metal_init: recommendedMaxWorkingSetSize  = 10922.67 MB
ggml_metal_init: hasUnifiedMemory              = true
ggml_metal_init: maxTransferRate               = built-in GPU
llama_new_context_with_model: compute buffer total size =   91.47 MB
llama_new_context_with_model: max tensor size =   128.17 MB
ggml_metal_add_buffer: allocated 'data            ' buffer, size =  7501.56 MB, ( 7502.12 / 10922.67)
ggml_metal_add_buffer: allocated 'eval            ' buffer, size =     1.48 MB, ( 7503.61 / 10922.67)
ggml_metal_add_buffer: allocated 'kv              ' buffer, size =   402.00 MB, ( 7905.61 / 10922.67)
ggml_metal_add_buffer: allocated 'alloc           ' buffer, size =    90.02 MB, ( 7995.62 / 10922.67)


 what is the capital of japan
The capital of Japan is Tokyo. [end of text]
0
ggml_metal_free: deallocating
```

### CODE REVIEW

File: ggml-metal.m [hunk 1]
```diff
@@ -140,12 +140,22 @@ @implementation GGMLMetalClass
 
     ctx->d_queue = dispatch_queue_create("llama.cpp", DISPATCH_QUEUE_CONCURRENT);
 
-#if 0
```
Comment (kchro3) [inline ggml-metal.m:143]: Thought it would be ok to replace this since it looked unfinished.

File: Package.swift [hunk 2]
```diff
@@ -4,23 +4,27 @@ import PackageDescription
...
-                "k_quants.c"
+                "k_quants.c",
+                "ggml-metal.m",
             ],
             publicHeadersPath: "spm-headers",
             cSettings: [
                 .unsafeFlags(["-Wno-shorten-64-to-32"]),
+                .unsafeFlags(["-fno-objc-arc"]),
+                .define("GGML_SWIFT"),
+                .define("GGML_USE_METAL"),
```
[the remark this answers is not in the document]
Comment (ggerganov) [reply Package.swift:27]: @kchro3 Let's address this comment and we can merge

File: Package.swift [hunk 3]
```diff
@@ -4,23 +4,27 @@ import PackageDescription
 
 let package = Package(
     name: "llama",
+    platforms: [.macOS(.v11)],
```
Comment (pkrmf) [inline Package.swift:7]: I haven't tried iOS but I do wonder if we are OK by limiting this package to macOS only. Can llamacpp run on iOS, tvOS or even watchOS @ggerganov?

File: Package.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (kchro3) [reply Package.swift]: how does it look now? i don't have an old macbook, but i was able to build it if i switched the if/else condition

File: Package.swift (hunk 3 above)
Comment (ggerganov) [reply Package.swift:21]: It can run even on a refrigerator 😄 

Jokes aside - I see no reason to limit this to just macOS

Comment (kchro3) [reply Package.swift:21]: i'm not sure why, but the compiler was not happy if platforms was not included, although that could be a metal thing?

Comment (kchro3) [reply Package.swift]: i don't have a way to test on watchOS or tvOS... i can set it to the minimum non-deprecated version and if someone tries to cross that bridge, we could update it?

Comment (pkrmf) [reply Package.swift]: do you mind sharing the compiler error?

Comment (kchro3) [reply Package.swift]: Screenshot 2023-09-08 at 11 16 37 AM

Comment (kchro3) [reply Package.swift]: too many errors, lol

Comment (jhen0409) [reply Package.swift]: https://github.com/mybigday/llama.rn/blob/main/scripts/ggml-metal.m.patch

I've a patch for iOS/tvOS, for skip some unavailable APIs. I also set the minimum versions to 11: [podspec](https://github.com/mybigday/llama.rn/blob/main/llama-rn.podspec#L24).

The .metal file should be able to skip compile by add to resources, it should be loaded dynamically during runtime.

Comment (pkrmf) [reply Package.swift]: Thats right. You shouldn’t be compiling the .metal file, it needs to be a copied resource for the spm target. That will cause your current code to access the bundle to have to change slightly since SPM generates a resources bundle with the metal file

Comment (kchro3) [reply Package.swift]: @jhen0409 , sorry do you mind giving me an ELI5? I'm a backend eng for the most part so not that familiar with Swift or React Native.

From what I can tell in the .m file, you're commenting out everything related to devices, but I'm not following why or if you're suggesting that I need to do something differently here.

RE: loading dynamically during runtime, when I imported llama through the Swift Package Manager, the .metallib was in the llama bundle. I think that the other part of the code (the else-clause of the GGML_SWIFT) is trying to load the file from source, but I'm not sure if that's necessary for Swift at least. Is there a reason to do that?

Comment (jhen0409) [reply Package.swift]: The API like MTLCopyAllDevices and some property in MTLDevice is not supported or requires newer version on iOS. I can open a PR and put more info for that.

> RE: loading dynamically during runtime, when I imported llama through the Swift Package Manager, the .metallib was in the llama bundle. I think that the other part of the code (the else-clause of the GGML_SWIFT) is trying to load the file from source, but I'm not sure if that's necessary for Swift at least. Is there a reason to do that?

I'm not sure what default.metallib used for. You can try use `resources: [ .copy(".metal") ]` in target & use `SWIFTPM_MODULE_BUNDLE` instead.

Comment (kchro3) [reply Package.swift]: Hey folks, I would appreciate help on this. I'm seeing in my build logs that it's compiling the .metal file even if I put it in `resources`.

For example, if I do:
```
#if arch(arm) || arch(arm64)
let platforms: [SupportedPlatform]? = [
    .macOS(.v11),
    .iOS(.v11),
    .watchOS(.v4),
    .tvOS(.v11)
]
let exclude: [String] = []
let resources: [Resource]? = [
    .copy("ggml-metal.metal"),
    .copy("README.md")  // just to validate that files get copied
]
let additionalSources: [String] = ["ggml-metal.m"]
let additionalSettings: [CSetting] = [
    .unsafeFlags(["-fno-objc-arc"]),
    .define("GGML_SWIFT"),
    .define("GGML_USE_METAL")
]
#else
```

I still see the default.metallib in my resources:
```
ls /Users/.../Library/Developer/Xcode/DerivedData/.../Build/Products/Debug/....app/Contents/Resources/llama_llama.bundle/Contents/Resources/
README.md		default.metallib
```


Screenshot 2023-09-08 at 6 37 22 PM


Could it be because the file is in the project root and the target path is "."?

Comment (pkrmf) [reply Package.swift]: what if you exclude ggml-metal.metal like it is done in master right now, but then add it in the resources section. That should do it I think

Comment (kchro3) [reply Package.swift]: https://developer.apple.com/documentation/packagedescription/target/exclude#discussion

I think it doesn't work because exclude takes precedence over resources. For example, I pushed a branch https://github.com/ggerganov/llama.cpp/pull/3091/files, and my build logs don't show the .metal file:

Screenshot 2023-09-08 at 7 57 54 PM

Comment (kchro3) [reply Package.swift]: https://developer.apple.com/library/archive/documentation/3DDrawing/Conceptual/MTLBestPracticesGuide/FunctionsandLibraries.html

Just a sanity check, but this documentation is saying that .metal files get automatically compiled, and that seems to be the case from what I've tried. Is there a specific reason why we need to compile from source?

Comment (jhen0409) [reply Package.swift]: > https://developer.apple.com/library/archive/documentation/3DDrawing/Conceptual/MTLBestPracticesGuide/FunctionsandLibraries.html
> 
> Just a sanity check, but this documentation is saying that .metal files get automatically compiled, and that seems to be the case from what I've tried.

Got it, it looks like a better way if we use it on Xcode.

**Comment by kchro3** [comment]
cc: @ggerganov, i think that all comments are now addressed

**Comment by jhen0409** [comment]
After I give it try, figured out the reason of the `ggml-metal.metal` compile errors on iOS:

- Required iOS / tvOS 14.0 to use `int64_t` in Metal ([ref](https://developer.apple.com/documentation/metal/mtldatatype/long))
- Required Metal >= v2.3 (Define MTL_LANGUAGE_REVISION = Metal23) ([ref](https://developer.apple.com/metal/Metal-Shading-Language-Specification.pdf): search `simd_sum`)

File: Package.swift [hunk 4]
```diff
@@ -2,32 +2,54 @@
 
 import PackageDescription
 
+#if arch(arm) || arch(arm64)
+let platforms: [SupportedPlatform]? = [
+    .macOS(.v11),
+    .iOS(.v11),
+    .watchOS(.v4),
+    .tvOS(.v11)
```
Comment (jhen0409) [inline Package.swift:10]:
```suggestion
    .macOS(.v11),
    .iOS(.v14),
    .watchOS(.v4),
    .tvOS(.v14)
```

File: Package.swift [hunk 5]
```diff
@@ -2,32 +2,54 @@
...
+]
+let exclude: [String] = []
+let additionalSources: [String] = ["ggml-metal.m"]
+let additionalSettings: [CSetting] = [
+    .unsafeFlags(["-fno-objc-arc"]),
+    .define("GGML_SWIFT"),
+    .define("GGML_USE_METAL")
+]
```
Comment (jhen0409) [inline Package.swift:18]:
```suggestion
    .define("GGML_SWIFT"),
    .define("GGML_USE_METAL"),
    .define("MTL_LANGUAGE_REVISION", to: "Metal23")
]
```

Comment (jhen0409) [reply Package.swift:18]: This is not necessary, ignore it.

**Review by ggerganov — APPROVED** [review]
@jhen0409 
Can't give this a test atm - if all works merge it
